### PR TITLE
Add elasticsearch implementation

### PR DESCRIPTION
**Pull request by danlamanna**

### CODE REVIEW

**Comment by danlamanna** [comment]
TODO before merge: filtering ages.

**Comment by danlamanna** [comment]
@brianhelba rebased and fixed up ages - ready for review.

**Comment by brianhelba** [comment]
https://aws.amazon.com/blogs/opensource/keeping-clients-of-opensearch-and-elasticsearch-compatible-with-open-source/ may be relevant here.

File: isic/core/models/image.py [hunk 1]
```diff
@@ -25,6 +25,21 @@ def __str__(self):
...
+        # The age has to be stored in the search index in rounded form, otherwise searches
+        # (e.g. 'age:47') could leak the true age.
+        if 'age' in m:
+            m['age'] = self.accession.rounded_age
+
+        for f in ['patient_id', 'lesion_id']:
+            if f in m:
+                del m[f]
+
+        return {**{'created': self.created, 'name': self.isic_id, 'public': self.public}, **m}
```
Comment (danlamanna) [inline isic/core/models/image.py:41]: name -> isic_id

File: isic/core/api.py [hunk 2]
```diff
@@ -29,3 +31,49 @@ class ImageViewSet(ReadOnlyModelViewSet):
...
+        query_serializer=SearchQuerySerializer,
+    )
+    @action(detail=False, methods=['get'], pagination_class=None)
+    def facets(self, request):
+        serializer = SearchQuerySerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+        query = {'bool': {}}
+        if not request.user.is_staff:
+            query['bool']['filter'] = {'term': {'public': 'true'}}
+
```
Comment (danlamanna) [inline isic/core/api.py:46]: code comment

Comment (danlamanna) [reply isic/core/api.py:46]: swap statements.

File: isic/core/api.py [hunk 3]
```diff
@@ -29,3 +31,49 @@ class ImageViewSet(ReadOnlyModelViewSet):
...
+    @action(detail=False, methods=['get'])
+    def search(self, request):
+        serializer = SearchQuerySerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+        query = {'bool': {}}
+        if not request.user.is_staff:
+            query['bool']['filter'] = {'term': {'public': 'true'}}
+
+        if serializer.data.get('query'):
+            query['bool']['must'] = {'query_string': {'query': serializer.data['query']}}
```
Comment (danlamanna) [inline isic/core/api.py:65]: helper method

File: isic/core/api.py [hunk 4]
```diff
@@ -29,3 +31,49 @@ class ImageViewSet(ReadOnlyModelViewSet):
...
+        )
+        isic_ids = [x['fields']['name'][0] for x in search_results['hits']['hits']]
+        images = self.get_queryset().filter(isic__id__in=isic_ids)
+        page = self.paginate_queryset(images)
+        serializer = self.get_serializer(page, many=True)
+        paginated_response = self.get_paginated_response(serializer.data)
+
+        # The count needs to be overriden otherwise the paginator tries to
+        # get a count for the queyrset (images), which will always be PAGE_SIZE.
+        paginated_response.data['count'] = search_results['hits']['total']['value']
```
Comment (danlamanna) [inline isic/core/api.py:78]: CountlessPaginator

File: isic/core/api.py [hunk 5]
```diff
@@ -29,3 +31,49 @@ class ImageViewSet(ReadOnlyModelViewSet):
...
+        if not request.user.is_staff:
+            query['bool']['filter'] = {'term': {'public': 'true'}}
+
+        if serializer.data.get('query'):
+            query['bool']['must'] = {'query_string': {'query': serializer.data['query']}}
+
+        search_results = search_images(
+            query, self.paginator.get_limit(request), self.paginator.get_offset(request)
+        )
+        isic_ids = [x['fields']['name'][0] for x in search_results['hits']['hits']]
```
Comment (danlamanna) [inline isic/core/api.py:70]: use pk

File: isic/core/tests/test_search.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+from isic.core.search import add_to_search_index, get_elasticsearch_client, maybe_create_index
+
+
+@pytest.fixture
+def search_index():
+    es = get_elasticsearch_client()
+    maybe_create_index()
+    yield
+    es.indices.delete(settings.ISIC_ELASTICSEARCH_INDEX)
```
Comment (danlamanna) [inline isic/core/tests/test_search.py:14]: is truncating faster?

File: isic/core/api.py [hunk 7]
```diff
@@ -19,10 +24,61 @@ def stats(request):
...
+    @swagger_auto_schema(
+        operation_description='Retrieve the facet counts of a query.',
+        query_serializer=SearchQuerySerializer,
+    )
+    @action(detail=False, methods=['get'], pagination_class=None)
+    def facets(self, request):
+        serializer = SearchQuerySerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+        query = build_filtered_query(
+            request.user, serializer.data['query'] if 'query' in serializer.data else None
```
Comment (brianhelba) [inline isic/core/api.py:57]:
```suggestion
            request.user, serializer.data.get('query')
```

File: isic/core/api.py [hunk 8]
```diff
@@ -19,10 +24,61 @@ def stats(request):
...
+    @swagger_auto_schema(
+        operation_description='Search images with a key:value query string.',
+        query_serializer=SearchQuerySerializer,
+    )
+    @action(detail=False, methods=['get'])
+    def search(self, request):
+        serializer = SearchQuerySerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+        query = build_filtered_query(
+            request.user, serializer.data['query'] if 'query' in serializer.data else None
```
Comment (brianhelba) [inline isic/core/api.py:70]:
```suggestion
            request.user, serializer.data.get('query')
```

File: isic/core/api.py [hunk 9]
```diff
@@ -19,10 +24,57 @@ def stats(request):
...
+    @swagger_auto_schema(
+        operation_description='Retrieve the facet counts of a query.',
+        query_serializer=SearchQuerySerializer,
+    )
+    @action(detail=False, methods=['get'], pagination_class=None)
+    def facets(self, request):
+        serializer = SearchQuerySerializer(data=request.query_params)
+        serializer.is_valid(raise_exception=True)
+        query = build_filtered_query(request.user, serializer.data.get('query'))
+        return JsonResponse(facets(query))
```
Comment (brianhelba) [inline isic/core/api.py:57]:
```suggestion
        return Response(facets(query))
```

File: isic/core/models/image.py [hunk 10]
```diff
@@ -25,6 +25,30 @@ def __str__(self):
     def get_absolute_url(self):
         return reverse('core/image-detail', args=[self.pk])
 
+    @property
+    def as_elasticsearch_document(self) -> dict:
+        m = self.accession.metadata
```
Comment (brianhelba) [inline isic/core/models/image.py:30]:
```suggestion
        m = dict(self.accession.metadata)
```

File: isic/core/search.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+def add_to_search_index(image) -> None:
+    get_elasticsearch_client().index(
+        index=settings.ISIC_ELASTICSEARCH_INDEX, body=image.as_elasticsearch_document
+    )
+
+
+def bulk_add_to_search_index(qs: QuerySet) -> None:
+    def image_documents():
+        for image in qs.select_related('accession').all():
+            yield image.as_elasticsearch_document
```
Comment (brianhelba) [inline isic/core/search.py:125]:
```suggestion
    image_documents = (
    	image.as_elasticsearch_document
    	for image in qs.select_related('accession').all()
  	)
```
Could this just be anything iterable?

**Review by brianhelba — APPROVED** [review]

**Pull request merged**